subsys/pm/device_runtime.c

```diff
@@ -282,19 +294,21 @@ int pm_device_runtime_disable(const struct device *dev)
 	SYS_PORT_TRACING_FUNC_ENTER(pm, device_runtime_disable, dev);
 
 	if (!k_is_pre_kernel()) {
-		(void)k_mutex_lock(&pm->lock, K_FOREVER);
+		key = k_spin_lock(&pm->lock);
 	}
 
 	if (!atomic_test_bit(&pm->flags, PM_DEVICE_FLAG_RUNTIME_ENABLED)) {
 		goto unlock;
 	}
 
 	/* wait until possible async suspend is completed */
-	if (!k_is_pre_kernel()) {
-		while (pm->state == PM_DEVICE_STATE_SUSPENDING) {
-			(void)k_condvar_wait(&pm->condvar, &pm->lock,
-					     K_FOREVER);
-		}
+	if (!k_is_pre_kernel() && pm->state == PM_DEVICE_STATE_SUSPENDING) {
+		k_spin_unlock(&pm->lock, key);
+
+		/* wait until possible async suspend is completed */
+		k_event_wait(&pm->event, EVENT_MASK, true, K_FOREVER);
+
+		key = k_spin_lock(&pm->lock);
```

Comment:
doesn't it introduces a possible race condition ?

Reply:
Can you expand on that?

Please note that this is doing the same of what `k_condvar_wait` was doing before but using spinlocks instead of mutexes (in the `k_condwar_wait` case the mutex was unlocked in the function itself, here we are doing that externally).